File: Leetcode/148. Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head:
            currNode, tailNode = head, None
            while currNode is not None:
                tailNode = currNode
                currNode = currNode.next
            return self.mergeSort(head, tailNode)
        return head

    def mergeSort(self, head, tail):
        if head == tail: return head
        currNode, length = head, 0
        while currNode is not None:
            length += 1
            currNode = currNode.next
        # To get to the middle node the middle node index is (1 + length) / 2 - 1
        length = ((1 + length) // 2) - 1
        currNode = head
        while length > 0:
            length -= 1
            currNode = currNode.next
        rightNode = currNode.next
        currNode.next = None
        return self.merge(self.mergeSort(head, currNode), self.mergeSort(rightNode, tail))

    def merge(self, L, R):
        if L and R:
            head = None
            if L.val < R.val:
                head = L
                L = L.next
            else:
                head = R
                R = R.next
            head.next = self.merge(L, R)
            return head
        elif L:
            return L
        else:
            return R

File: Leetcode/148. Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode findMidNode(ListNode head) {
        ListNode midNode = head;
        while (head != null && head.next != null && head.next.next != null) {
            head = head.next.next;
            midNode = midNode.next;
        }
        return midNode;
    }
    private ListNode mergeLists(ListNode A, ListNode B) {
        if (A != null && B != null) {
            if (A.val > B.val) return mergeLists(B, A);
            ListNode head = A;
            head.next = mergeLists(A.next, B);
            return head;
        }
        if (A != null) return A;
        return B;
    }
    private ListNode mergeSort(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode mid = findMidNode(head);
        if (mid != null) {
            ListNode rightNode = mid.next;
            mid.next = null;
            return mergeLists(mergeSort(head), mergeSort(rightNode));
        }
        return mid;
    }
    public ListNode sortList(ListNode head) {
        return mergeSort(head);
    }
}